vue2/src/views/EventListWithCompositionApi.vue

```diff
@@ -0,0 +1,52 @@
+<template>
+  <div v-if="user && user.user">
+    <h1>Events for {{ user.user.name }}</h1>
+    <Events :page="page" :perPage="perPage"/>
+    <template v-if="page != 1">
+      <router-link
+        :to="{ name: 'event-list-with-composition-api', query: { page: page - 1 } }"
+        rel="prev"
+      >
+        Prev Page
+      </router-link>
+    </template>
+    <template v-if="page < lastPage">
+      <router-link
+        :to="{ name: 'event-list-with-composition-api', query: { page: page + 1 } }"
+        rel="next"
+      >
+        Next Page
+      </router-link>
+    </template>
+  </div>
+</template>
+
+<script>
+import {
+  ref,
+} from '@vue/composition-api';
+import Events from '@/components/EventsWithCompositionApi.vue';
+import useUserEvents from '@/composables/useUserEvents';
+import { mapState } from 'vuex';
+
+export default {
+  components: {
+    Events,
+  },
+  setup(props, context) {
+    // TODO $route.query는 vue-router4(Vue3)에서는 useRoute으로 바꿔야 함.
+    const page = ref(parseInt(context.root.$route.query.page, 10) || 1);
```

Comment:
```suggestion
    const page = computed(() => parseInt(context.root.$route.query.page, 10) || 1);
```
기존 코드를 그대로 옮겨 온다면, 요 아이는 동일하게 computed로 해야 문맥이 통하지 않을까 싶습니다.

근데, 결국 composition api로 생성하면, computed로 만들어지는게 아니라, data로 만들어지네요. 🤔

Reply:
@KunHwanAhn 좋은 제안 감사드려요. `page`란 데이터는 브라우저에서 입력한 queryString 인자를 그대로 쓰는 흐름이므로, vue2 legacy에서 이를 data 메서드에서 받아서 쓰는 형태로 변경하고 이곳은 ref를 그대로 쓰는 것이 어떨까합니다. 확인부탁드려요. (computed는 원본 데이터를 다른 성격으로 가공하는 경우에 쓰는 것이 더 적합하다고 생각해서요.)

제가 참고한 자료: 
`That's why for complex logic that includes reactive data, you should use a computed property.`
[Vue3 Computed Properties and Wathcers](https://v3.vuejs.org/guide/computed.html)

```
// 그러므로 EventList.vue만 수정하면 될 것 같습니다.
  data() {
    return {
      perPage: 3,
      page: parseInt(this.$route.query.page, 10) || 1,
    };
  },
```

Reply:
아! 여기서의 요지는 이전 파일에서 `computed로 사용하고 있었으니, 여기서도 동일하게 computed를 사용`하는 것 이었습니다!

그리고, `$route.query.page`에 대해서는 직접 $route 에 접근하는 것도 괜찮지만, Router의 `props`에서 VueRouter에서 지정하는 컴포넌트의 Props로 전달하는 방향을 쓰는 것을 제안해봅니다! 🙏 

```javascript
  {
      path: '/event-list-with-composition-api',
      name: 'event-list-with-composition-api',
      component: EventListWithCompositionApi,
      props: (route) => {
        const { page } = route.query;

        return {
          query: {
            page: parseInt(page, 10) || 1,
          },
        };
      },
      meta: {
        requiresAuth: true,
      },
    },
```

Reply:
음.. 그리고 네트워크 탭을 확인해보니, API를 2번씩 호출하고 있네요. 🤔 

![스크린샷 2021-02-03 오후 9 13 33](https://user-images.githubusercontent.com/3446660/106745706-d74f6700-6664-11eb-9ee0-73f8c61aef57.png)

EventListWithCompositionApi / EventsWithCompositionApi 2곳에서 `useUserEvents`를 사용하고 있어서, 두 컴포넌트의 mounted 훅에서 각각 1번씩 2번 호출하는 것 같습니다. 🤔  이건 확인 부탁드릴게요! 🙏

Reply:
> 아! 여기서의 요지는 이전 파일에서 `computed로 사용하고 있었으니, 여기서도 동일하게 computed를 사용`하는 것 이었습니다!
> 
> 그리고, `$route.query.page`에 대해서는 직접 $route 에 접근하는 것도 괜찮지만, Router의 `props`에서 VueRouter에서 지정하는 컴포넌트의 Props로 전달하는 방향을 쓰는 것을 제안해봅니다! 🙏
> 
> ```js
>   {
>       path: '/event-list-with-composition-api',
>       name: 'event-list-with-composition-api',
>       component: EventListWithCompositionApi,
>       props: (route) => {
>         const { page } = route.query;
> 
>         return {
>           query: {
>             page: parseInt(page, 10) || 1,
>           },
>         };
>       },
>       meta: {
>         requiresAuth: true,
>       },
>     },
> ```

이 수정 사항은 흥미롭네요. 말씀주신대로 적용을 하게 되면 setup 내부에서는 router에서 전달한 props 정보를 접근하지 못하네요. 하지만 props를 setup 외부에 선언하면 참조할 수 있습니다.

```js
  // FIX ME Vue2에서는 Router에서 전달한 props 정보는 setup에서 직접 받을 수 없습니다.
  // props를 선언하고 받을 parameter를 정의해야 setup에서 참조할 수 있습니다.
  props: { // 이 props를 추가해야 합니다.
    query: {
      type: Object,
      required: true,
    },
  },
  setup(props) {
    const page = ref(props.query.page);
    const perPage = ref(3);
    const { lastPage } = useUserEvents(perPage, page);

    return {
      page,
      perPage,
      lastPage,
    };
  },

```

Reply:
넵 맞습니다. 위의 props 블록 안에서 query 을 선언해줘야 합니다. 🙏

Reply:
@KunHwanAhn Api를 2번씩 호출하는 이슈에 대해 확인해보니 아래 문제점을 발견했습니다.

- composable의 useUserEvent는 import 하는 곳마다 데이터를 별도로 보관하고 있다.(그래서 한번 fetch해도 데이터를 공유하지 못한다.)

이를 해결하기 위해서 다음과 같은 조치를 취했습니다.
```js
// useUserEvent.js
// 1. 반응형 참조를 함수 바깥에 선언해서 같은 반응형 참조를 바라보도록 합니다.
const events = ref([]);
const totalCount = ref(0);
const lastPage = ref(0);

export default function useUserEvents(perPage, page) {
   // 중략...
}

// 2. componentApi를 Vue.use() 메서드에 설정하는 시점을 App.vue를 로딩하기 이전에 완료되도록 처리합니다.
// componentApi/index.js
```
이 조치로 Api 1회 호출 및 비슷한 데이터를 가져다 써야하는 컴포넌트들에게 Vuex만이 아닌 composable module을 선택할 수 있다는 유연성을 줄 수 있었습니다.

이런 형태가 Vuex를 기존에 사용하던 목적에 영향을 준다고 생각합니다. 어느정도 composable안의 모듈로만 해결이 가능한 범위의 데이터를 다룬다면, Vuex의 global state에 저장해두지 않아도 활용할 수 있겠다는 판단이 들었습니다. 

아울러 2번의 해결방안은 특정 가이드가 아닌 에러 메시지의 안내에 따라 진행되었습니다. 이 부분도 잠시 의견 나누면 좋겠습니다.

Reply:
```vue.js
// useUserEvent.js
// 1. 반응형 참조를 함수 바깥에 선언해서 같은 반응형 참조를 바라보도록 합니다.
const events = ref([]);
const totalCount = ref(0);
const lastPage = ref(0);
```

이렇게 쓰면 요 3변수는 모두 모듈 변수가 되어서, 모든 컴포넌트에서 공유하는 녀석이 되어버려서... 조금 애매하긴 하네요...

이건 구두로 논의 했던 것처럼, `SPA의 공통 상태`에 저장해서 관리할 데이터가 아니라고 판단했기에... 요녀석은 https://github.com/wonder13662/migration-vue2-vue3/issues/23 이슈 수정하면서 같이 논의해봐요!